fe/model.py

```diff
@@ -81,23 +203,51 @@ def predict(self, ff_params: list, mol_a: Chem.Mol, mol_b: Chem.Mol, core: np.nd
-            # the hysteresis by jointly minimizing against a and b at the same time. We may also want
-            # to remove the randomness completely from the minimization.
-            min_host_coords = minimizer.minimize_host_4d([mol_a, mol_b], host_system, host_coords, self.ff, host_box)
-
             single_topology = topology.SingleTopology(mol_a, mol_b, core, self.ff)
             rfe = free_energy.RelativeFreeEnergy(single_topology)
-
-            unbound_potentials, sys_params, masses, coords = rfe.prepare_host_edge(ff_params, host_system, min_host_coords)
+            edge_hash = self._edge_hash(stage, mol_a, mol_b, core)
+            if self.pre_equilibrate and edge_hash in self._equil_cache:
+                x0, _, host_box = self._equil_cache[edge_hash]
+                num_host_coords = len(host_coords)
+                unbound_potentials, sys_params, masses, _ = rfe.prepare_host_edge(ff_params, host_system, host_coords)
+                mol_a_size = mol_a.GetNumAtoms()
+                # Use Dual Topology to pre equilibrate, so have to get the mean of the two sets of mol,
+                # normally done within prepare_host_edge, but the whole system has moved by this stage
+                x0 = np.concatenate(
+                    [
+                        x0[:num_host_coords],
+                        np.mean(
+                            single_topology.interpolate_params(
+                                x0[num_host_coords:num_host_coords+mol_a_size],
+                                x0[num_host_coords+mol_a_size:]
+                            ),
+                            axis=0
+                        )
+                    ]
+                )
+            else:
+                if self.pre_equilibrate:
+                    print("Edge not correctly pre-equilibrated, ensure equilibrate_edges was called")
+                print(f"Minimizing the {stage} host structure to remove clashes.")
+                # (ytz): this isn't strictly symmetric, and we should modify minimize later on remove
+                # the hysteresis by jointly minimizing against a and b at the same time. We may also want
+                # to remove the randomness completely from the minimization.
+                min_host_coords = minimizer.minimize_host_4d([mol_a, mol_b], host_system, host_coords, self.ff, host_box)
+
+                unbound_potentials, sys_params, masses, coords = rfe.prepare_host_edge(ff_params, host_system, min_host_coords)
+
+                x0 = coords
+            v0 = np.zeros_like(x0)
+
+
+            time_step = 1.5e-3
 
             harmonic_bond_potential = unbound_potentials[0]
-            group_idxs = get_group_indices(get_bond_list(harmonic_bond_potential))
-
-            x0 = coords
-            v0 = np.zeros_like(coords)
+            bond_list = get_bond_list(harmonic_bond_potential)
+            if self.hmr:
+                masses = apply_hmr(masses, bond_list)
+                time_step = 2.5e-3
```

Comment:
I don't see a much better alternative, but I wonder if there's a way to avoid `timestep = literal; if self.hmr: timestep = different_literal`. Perhaps we want to move the responsibility for changing the masses into `rfe.prepare_host_edge(..., hmr=True)`, and make `time_step` an explicit argument in the RBFEModel constructor.

Reply:
We're going to refactor these classes like crazy soon with all the dual topology stuff soon to come.